core/arch/arm/plat-zynqmp/sta_bsload.c

```diff
@@ -0,0 +1,366 @@
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pcap_init(void)
+{
+	TEE_Result ret;
+
+	/* reset PCAP */
+	write32(CSU_PCAP_RESET_RESET, CSU_PCAP_RESET);
+	write32(0, CSU_PCAP_RESET);
+
+	write32(CSU_PCAP_CTRL_PCAP_PR, CSU_PCAP_CTRL);
+	write32(0, CSU_PCAP_RDWR);
+
+	/* reset PL */
+	write32(0, CSU_PCAP_PROG);
+	ret = poll_reg_tout_32(CSU_PCAP_STATUS, CSU_PCAP_STATUS_PL_RESET_B, 0,
+			       0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout waiting for PL reset\n");
+		return ret;
+	}
+
+	write32(CSU_PCAP_PROG_PCFG_PROG_B, CSU_PCAP_PROG);
+	ret = poll_reg_tout_32(CSU_PCAP_STATUS, CSU_PCAP_STATUS_PL_INIT,
+			       CSU_PCAP_STATUS_PL_INIT, 0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout waiting for PL init done\n");
+		return ret;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result pcap_write(uint64_t srcaddr, uint32_t size)
+{
+	TEE_Result ret;
+	uint32_t regval;
+
+	write32(0, CSUDMA_SRC_CRC);
+
+	regval = read32(CSU_SSS_CONFIG);
+	regval &= CSU_SSS_CFG_PCAP_SSS_MASK;
+	regval |= (CSU_SSS_CFG_PCAP_DMA_DATA << CSU_SSS_CFG_PCAP_SSS_SHFT);
+	write32(regval, CSU_SSS_CONFIG);
+
+	write32((uint32_t)srcaddr, CSUDMA_SRC_ADDR);
+	write32((srcaddr >> 32) & 0x1FFFF, CSUDMA_SRC_ADDR_MSB);
+	write32(size & CSUDMA_SRC_SIZE_MASK, CSUDMA_SRC_SIZE);
+
+	ret = poll_reg_tout_32(CSUDMA_SRC_I_STS, CSUDMA_SRC_I_STS_DONE,
+			       CSUDMA_SRC_I_STS_DONE, 0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout writing bitstream\n");
+		return ret;
+	}
+
+	write32(CSUDMA_SRC_I_STS_DONE, CSUDMA_SRC_I_STS);
+
+	ret = poll_reg_tout_32(CSU_PCAP_STATUS, CSU_PCAP_STATUS_PCAP_WR_IDLE,
+			       CSU_PCAP_STATUS_PCAP_WR_IDLE, 0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout waiting for CSUDMA idle\n");
+		return ret;
+	}
+
+	DMSG("SRC_CRC: %#x\n", read32(CSUDMA_SRC_CRC));
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result wait_done(void)
+{
+	TEE_Result ret;
+	uint32_t regval;
+
+	ret = poll_reg_tout_32(CSU_PCAP_STATUS, CSU_PCAP_STATUS_PL_DONE,
+			       CSU_PCAP_STATUS_PL_DONE, 0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout waiting for PL done\n");
+		return ret;
+	}
+
+	regval = read32(CSU_PCAP_RESET);
+	regval |= CSU_PCAP_RESET_RESET;
+	write32(regval, CSU_PCAP_RESET);
+	ret = poll_reg_tout_32(CSU_PCAP_RESET, CSU_PCAP_RESET_RESET,
+			       CSU_PCAP_RESET_RESET, 0x8000000);
+	if (ret != TEE_SUCCESS) {
+		DMSG("timeout waiting for PCAP reset assert\n");
+		return ret;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result do_load_bs(vaddr_t bs, size_t sz, uint64_t flags)
+{
+	paddr_t addr;
+	TEE_Result ret;
+	(void)flags;
+
+	/* sanity check bitstream */
+	if (bs & 0x3)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (sz > BS_SIZE_MAX)
+		return TEE_ERROR_EXCESS_DATA;
+
+	ret = check_header(bs);
+	if (ret)
+		return ret;
+
+	addr = virt_to_phys((void *)bs);
+	if (!addr) {
+		DMSG("no mapping found for va '%#lx'\n", bs);
+		return TEE_ERROR_GENERIC;
+	}
+	DMSG("bitstream va:%#lx, pa:%#lx, sz:%#zx\n", bs, addr, sz);
+
+	flush_dcache_range(bs, sz);
```

Comment:
The name of the function is misleading, it handles the internal caches. `flush_dcache_range()` is only available in Aarch64.